patches/api/0332-Update-Folder-Uses-Plugin-Name.patch

```diff
@@ -0,0 +1,86 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: samue <xemoryt@gmail.com>
+Date: Tue, 7 Sep 2021 18:11:37 +0100
+Subject: [PATCH] Update-Folder-Uses-Plugin-Name
+
+
+diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+index 0e25119564dfa9cb12f3c5dc5f653d7f2c147a9d..3e3ec3de79b0ea4ed99ea354403ccc92510f60f1 100644
+--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
++++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+@@ -399,7 +399,7 @@ public final class SimplePluginManager implements PluginManager {
+     public synchronized Plugin loadPlugin(@NotNull File file) throws InvalidPluginException, UnknownDependencyException {
+         Validate.notNull(file, "File cannot be null");
+ 
+-        checkUpdate(file);
++        file = checkUpdate(file); // Paper - update the reference in case checkUpdate renamed it
+ 
+         Set<Pattern> filters = fileAssociations.keySet();
+         Plugin result = null;
+@@ -410,7 +410,6 @@ public final class SimplePluginManager implements PluginManager {
+ 
+             if (match.find()) {
+                 PluginLoader loader = fileAssociations.get(filter);
+-
+                 result = loader.loadPlugin(file);
+             }
+         }
+@@ -426,16 +425,53 @@ public final class SimplePluginManager implements PluginManager {
+         return result;
+     }
+ 
+-    private void checkUpdate(@NotNull File file) {
++    // Paper start - Update Folder Uses Plugin Name to replace
++    /**
++     * Replaces a plugin with a plugin of the same plugin name in the update folder.
++     * @param file
++     * @throws InvalidPluginException
++     */
++    private File checkUpdate(@NotNull File file) throws InvalidPluginException {
+         if (updateDirectory == null || !updateDirectory.isDirectory()) {
+-            return;
++            return file;
++        }
++        PluginLoader pluginLoader = getPluginLoader(file);
++        try {
++            String pluginName = pluginLoader.getPluginDescription(file).getName();
++            for (File updateFile : updateDirectory.listFiles()) {
++                if (updateFile.isFile()) {
++                    PluginLoader updatePluginLoader = getPluginLoader(updateFile);
```

Comment:
Probably a good idea to cache this so we don't get the plugin loader every time.

Reply:
Yeah I was thinking that but 
1) Not sure where I'd do the caching exactly / how I'd handle that
2) I don't think it's much of an issue in practice, on my network I'm usually only updating like four plugins max at a time.

Reply:
Lynx and I talked about this for a bit, not having caching here is probably fine and easier since in most use cases only a couple of plugins will be updated a time.

Reply:
yeah that's what I was saying as my second point